lucene/core/src/java/org/apache/lucene/search/FilteringFieldComparator.java

```diff
@@ -0,0 +1,350 @@
+        private int maxDocVisited;
+        private int updateCounter = 0;
+        private final String field;
+        protected boolean canUpdateIterator = false; // set to true when queue becomes full and hitsThreshold is reached
+
+        public FilteringNumericComparator(NumericComparator<T> in, boolean reverse, int bytesCount) {
+            super(in);
+            this.field = in.field;
+            this.bytesCount = bytesCount;
+            this.reverse = reverse;
+            minValueAsBytes = new byte[bytesCount];
+            maxValueAsBytes = new byte[bytesCount];
+            if (reverse) {
+                minValueExist = true;
+            } else {
+                maxValueExist = true;
+            }
+        }
+
+        @Override
+        public void setCanUpdateIterator() throws IOException {
+            this.canUpdateIterator = true;
+            // for the 1st time queue becomes full and hitsThreshold is reached
+            // we can start updating competitive iterator
+            updateCompetitiveIterator();
+        }
+
+        @Override
+        public void setTopValue(T value) {
+            hasTopValue = true;
+            if (reverse) {
+                maxValueExist = true;
+            } else {
+                minValueExist = true;
+            }
+            in.setTopValue(value);
+        }
+
+        @Override
+        public void setBottom(int slot) throws IOException {
+            ((NumericComparator) in).setBottom(slot);
+            updateCompetitiveIterator(); // update an iterator if we set a new bottom
+        }
+
+        @Override
+        public int compareBottom(int doc) throws IOException {
+            return ((NumericComparator) in).compareBottom(doc);
+        }
+
+        @Override
+        public int compareTop(int doc) throws IOException {
+            return ((NumericComparator) in).compareTop(doc);
+        }
+
+        @Override
+        public void copy(int slot, int doc) throws IOException {
+            ((NumericComparator) in).copy(slot, doc);
+        }
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {}
+
+        @Override
+        public final LeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException {
+            ((NumericComparator) in).doSetNextReader(context);
+            pointValues = context.reader().getPointValues(field);
+            iterator = pointValues == null ? null : ((NumericComparator)in).currentReaderValues;
+            maxDoc = context.reader().maxDoc();
+            maxDocVisited = 0;
+            updateCompetitiveIterator(); // update an iterator if we have a new segment
+            return this;
+        }
+
+        // update its iterator to include possibly only docs that are "stronger" than the current bottom entry
+        public void updateCompetitiveIterator() throws IOException {
+            if (canUpdateIterator == false) return;
+            if (pointValues == null) return;
```

Comment:
I believe that this `pointValues == null` aims at disabling the optimization when the field has doc values but is not indexed. However it can also be null if the field is indexed but doesn't have values on this segment. It's fine to not optimize this case right now, but let's at least leave a comment about it?

Reply:
addressed in  24c94ff